airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py

```diff
@@ -0,0 +1,268 @@
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        yield self._extend_record(self._api.account, fields=fields)
+
+    def _extend_record(self, ig_user, fields) -> Dict:
+        return ig_user.api_get(fields=fields).export_all_data()
+
+
+class UserLifetimeInsightsAPI(StreamAPI):
+    LIFETIME_METRICS = ["audience_city", "audience_country", "audience_gender_age", "audience_locale"]
+    period = "lifetime"
+
+    buffer_days = 29
+    days_increment = 1
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        account_id = self._api.account.get("id")
+        for insight in self._get_insight_records(params=self._params()):
+            yield {
+                "id": account_id,
+                "metric": insight.get("name"),
+                "date": insight.get("values")[0]["end_time"],
+                "value": insight.get("values")[0]["value"],
+            }
+
+    def _params(self) -> Dict:
+        return {"metric": self.LIFETIME_METRICS, "period": self.period}
+
+    def _get_insight_records(self, params) -> Iterator[Any]:
+        return self._api.account.get_insights(params=params)
+
+
+class UserInsightsAPI(IncrementalStreamAPI):
+    METRICS_BY_PERIOD = {
+        "day": [
+            "email_contacts",
+            "follower_count",
+            "get_directions_clicks",
+            "impressions",
+            "phone_call_clicks",
+            "profile_views",
+            "reach",
+            "text_message_clicks",
+            "website_clicks",
+        ],
+        "week": ["impressions", "reach"],
+        "days_28": ["impressions", "reach"],
+        "lifetime": ["online_followers"],
+    }
+
+    state_pk = "date"
+
+    buffer_days = 29
+    days_increment = 1
+
+    def __init__(self, api):
+        super().__init__(api=api)
+        self._state = api._start_date
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        account_id = self._api.account.get("id")
+        for params_per_day in self._params():
+            insight_list = []
+            for params in params_per_day:
+                insight_list += self._get_insight_records(params=params)
+            if not insight_list:
+                continue
+
+            insight_record = {"id": account_id}
+            for insight in insight_list:
+                key = (
+                    f"{insight.get('name')}_{insight.get('period')}"
+                    if insight.get("period") in ["week", "days_28"]
+                    else insight.get("name")
+                )
+                insight_record[key] = insight.get("values")[0]["value"]
+                if not insight_record.get("date"):
+                    insight_record["date"] = insight.get("values")[0]["end_time"]
+            yield self.state_filter(insight_record)
+
+    def _params(self) -> Iterator[List]:
+        buffered_start_date = max(self._state.add(minutes=1), pendulum.now().subtract(days=self.buffer_days))
```

Comment:
is insight data retained only for `buffer_days`? what if I want insight data from 1 year ago?

Reply:
Unfortunately this is not possible. While trying to do this - I get the following error from FB.
```
{
  "error": {
    "message": "(#100) There cannot be more than 30 days (2592000 s) between since and until.",
    "type": "OAuthException",
    "code": 100,
    "fbtrace_id": "As4kde-yMtu4oXtUCBMH6_T"
  }
}
```

2592000s = 30 days (today and 29 days before)

Reply:
👌🏼

Reply:
can you add this as a comment also? (That retention period is 29 days)

Reply:
@yevhenii-ldv but doesn't it mean that I can have start and end really old, but only small difference between them?

so I can read old data with multiple calls, 

- 2001-01-01 - 2001-02-01
- 2001-02-01 - 2001-03-01
- ...